src/lazy/expanded/macro_evaluator.rs

```diff
@@ -761,6 +1091,235 @@ mod tests {
         )
     }
 
+    mod cardinality {
+
+        mod bang {
+            use crate::lazy::expanded::macro_evaluator::tests::eval_template_invocation;
+
+            #[test]
+            #[should_panic]
+            fn required_does_not_accept_empty_rest() {
+                eval_template_invocation(
+                    "(macro foo (x) (make_string x x))",
+                    r#"
+                (:foo)
+            "#,
+                    r#"
+                // should raise an error
+            "#,
+                )
+                .unwrap()
+            }
+
+            #[test]
+            #[should_panic]
+            fn required_does_not_accept_empty_arg_group() {
+                eval_template_invocation(
+                    "(macro foo (x) (make_string x x))",
+                    r#"
+                (:foo (:))
+            "#,
+                    r#"
+                // should raise an error
+            "#,
+                )
+                .unwrap()
+            }
+
+            #[test]
+            #[should_panic]
+            fn required_does_not_accept_populated_arg_group() {
+                eval_template_invocation(
+                    "(macro foo (x) (make_string x x))",
+                    r#"
+                (:foo (:))
+            "#,
+                    r#"
+                // should raise an error
+            "#,
+                )
+                .unwrap()
+            }
+        }
+
+        mod optional {
+            use crate::lazy::expanded::macro_evaluator::tests::eval_template_invocation;
+            use crate::IonResult;
+
+            #[test]
+            fn optional_accepts_empty_or_expr() -> IonResult<()> {
+                eval_template_invocation(
+                    "(macro foo (x?) (make_string x x))",
+                    r#"
+                (:foo)           // x is implicitly empty
+                (:foo (:))       // x is explicitly empty
+                (:foo (:   ))    // x is explicitly empty with extra whitespace
+                (:foo "a")       // x is "a"
+                (:foo (:foo a))  // x is `(:foo a)`
+            "#,
+                    r#"
+                ""
+                ""
+                ""
+                "aa"
+                "aaaa"
+            "#,
+                )
+            }
+
+            #[test]
+            fn narrow() -> IonResult<()> {
```

Comment:
Can you make this name more descriptive and/or add a comment?

Reply:
I have no idea what that name referred to. x_x

I'm pretty sure it was a one-off to debug something. I've removed it.